<a href="https://colab.research.google.com/github/AxelMarcelino/AxelMarcelino/blob/main/PrakML1_Axel_Marcelino_Sondakh.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Axel Marcelino Sondakh**

**24060120140166**

# Understanding Dataset

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_excel('/content/drive/MyDrive/PrakML/Cryotherapy.xlsx')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Melakukan koneksi dengan Google Drive yang digunakan untuk menyimpan dataset, yaitu dataset Cryotherapy.xlsx di folder PrakML dengan fitur drive dari library google.colab dan fitur pd.read_excel dari library pandas untuk mengambil dataset tersebut dari Google Drive dan kita gunakan untuk proses Machine Learning pada tugas praktikum kali ini. Dataset tersebut saya namai dengan df.

Dataset Cryotherapy.xlsx ini saya ambil dari website UCI Machine Learning, yaitu https://archive.ics.uci.edu/ml/datasets/Cryotherapy+Dataset+#

In [ ]:
df.head(10)

sex  age   Time  Number_of_Warts  Type  Area  Result_of_Treatment
0    1   35  12.00                5     1   100                    0
1    1   29   7.00                5     1    96                    1
2    1   50   8.00                1     3   132                    0
3    1   32  11.75                7     3   750                    0
4    1   67   9.25                1     1    42                    0
5    1   41   8.00                2     2    20                    1
6    1   36  11.00                2     1     8                    0
7    1   59   3.50                3     3    20                    0
8    1   20   4.50               12     1     6                    1
9    2   34  11.25                3     3   150                    0

Sedikit tampilan mengenai dataset Cryotherapy.xlsx sebanyak 10 baris

# Splitting Dataset
### Memisahkan dataset menjadi Training dan Testing Dataset dengan rasio 80% untuk Training Dataset dan 20% untuk Testing Dataset

In [ ]:
from sklearn import model_selection

# Dependent Variable
y = df["Result_of_Treatment"]
# Independent Variable
X = df.drop(["Result_of_Treatment"], axis = 1)

X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size = 0.2, random_state = 7)

Dilakukan pemisahan dataset menjadi Training dan Testing (Validation) Dataset dengan rasio 80% untuk Training Dataset dan 20% untuk Testing (Validation) Dataset dengan salah satu fitur library Scikit-Learn, yaitu train_test_split dengan random_state = 7

Di bawah ini merupakan shape dari Training dan Testing Dataset untuk memperlihatkan secara lebih jelas / detail mengenai seberapa banyak baris dan kolom yang dimiliki masing-masing Training dan Testing Dataset. X_train merupakan Independent Variables untuk Training Dataset, y_train merupakan Dependent Variables untuk Training Dataset, X_validation merupakan Independent Variables untuk Testing (Validation) Dataset, dan y_validation merupakan Dependent Variables untuk Testing (Validation) Dataset.

In [ ]:
X_train.shape

(72, 6)

In [ ]:
y_train.shape

(72,)

In [ ]:
X_validation.shape

(18, 6)

In [ ]:
y_validation.shape

(18,)

# Algorithm Model Evaluation
### Memilih Model Algoritma Terbaik berdasarkan nilai mean dan standar deviasi

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

Mengimpor fitur KNeighborsClassifier (KNN), GaussianNB (Gaussian Naive Bayes), dan SVC (Support Vector Machine / SVM) untuk digunakan sebagai model dalam tugas praktikum Machine Learning kali ini dari library Scikit-Learn (sklearn.neighbors, sklearn.naive_bayes, dan sklearn.svm)

In [ ]:
models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

Memasukkan seluruh model (KNeighborsClassifier, GaussianNB, dan SVC) ke dalam array yang diberi nama models

In [ ]:
results = []
names = []

for (name, model) in models:
  kfold = KFold(n_splits=10, random_state=7, shuffle = True)
  cv_results = cross_val_score(model, X_train, y_train, cv = kfold, scoring= 'accuracy')
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" %(name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.719643 (0.158527)
NB: 0.844643 (0.149414)
SVM: 0.482143 (0.199489)


Mencari nilai rata-rata (mean) dan standar deviasi (std) dari 3 model tersebut untuk menentukan model algoritma yang terbaik untuk dataset Cryotherapy.xlsx ini yang di mana semakin tinggi nilai mean dan semakin rendah nilai standar deviasi, maka semakin baik pula model tersebut untuk dataset kita. Pencarian nilai mean dan std tersebut menggunakan bantuan fitur KFold() untuk membuat Folds dengan sebanyak 10 Folds, cross_val_score() untuk mengembalikan nilai berupa array yang terdiri dari akurasi pengujian setiap fold dari dataset, mean() untuk mencari nilai mean (rata-rata), dan std() untuk mencari nilai standar deviasi. Pada tahap ini, dapat kita lihat bahwa model algoritma yang terbaik merupakan model Gaussian Naive Bayes dengan nilai mean sebanyak 0.84463 dan nilai standar deviasi sebanyak 0.149414.

# Gaussian Naive Bayes Evaluation
### Dilakukan evaluasi dengan melihat nilai accuracy dan juga melihat confusion matrix serta classification report untuk melihat hasil evaluasi lainnya, seperti nilai precision, recall, dan f1 score dari dataset kita

In [ ]:
from sklearn import metrics

GNB = GaussianNB()
GNB = GNB.fit(X_train, y_train)
y_pred = GNB.predict(X_validation)

print("Accuracy Score Gaussian Naive Bayes : ", accuracy_score(y_validation, y_pred), "\n")
print("Confusion Matrix :\n", confusion_matrix(y_validation, y_pred), "\n")
print("Classification Report :\n", classification_report(y_validation, y_pred))

Accuracy Score Gaussian Naive Bayes :  0.8888888888888888 

Confusion Matrix :
 [[ 4  2]
 [ 0 12]] 

Classification Report :
               precision    recall  f1-score   support

           0       1.00      0.67      0.80         6
           1       0.86      1.00      0.92        12

    accuracy                           0.89        18
   macro avg       0.93      0.83      0.86        18
weighted avg       0.90      0.89      0.88        18



Pada tahap terakhir ini, kita melakukan evaluasi terhadap model algoritma yang kita pilih, yaitu Gaussian Naive Bayes. Di sini, kita akan dapat melihat nilai accuracy dari model algoritma Gaussian Naive Bayes yaitu 89%, Confusion Matrix-nya, dan Classification Report-nya yang di mana kita dapat melihat nilai precision, recall, f1-score, dan yang lainnya.